homeassistant/components/mqtt/mixins.py

```diff
@@ -523,6 +523,11 @@ async def discovery_callback(payload):
     async def async_removed_from_registry(self) -> None:
         """Clear retained discovery topic in broker."""
         if not self._removed_from_hass:
+            # Stop subscribing to discovery updates to not trigger when we clear the
+            # discovery topic
+            self._cleanup_discovery_on_remove()
+
+            # Clear the discovery topic so the entity is not rediscovered after a restart
```

Comment:
FTR: This fixes an old bug which surfaced due to changed timing when `MqttSensor` and `MqttBinarySensor` extend `RestoreEntity`

Reply:
The bug fix also applies for other platforms like MqttSwitch